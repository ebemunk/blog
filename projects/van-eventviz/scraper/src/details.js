import * as R from 'ramda'
import WAE from 'web-auto-extractor'
import axios from 'axios'
import Promise from 'bluebird'

import { logger, qs } from './util'
import { getPool } from './db'

const log = logger({
  slack: {
    username: 'Details Dude',
    icon_emoji: ':japanese_goblin:',
  },
})

export default async function details(args, opts) {
  log.verbose({
    type: 'env',
    env: process.env,
  })
  const pool = getPool()
  const { rows: list } = await pool.query(
    `select * from events where details is null or (geo is null and details!='false'::jsonb);`,
    // `select * from events where id='1027356';`,
  )
  log.verbose({
    type: 'found events without details',
    count: list.length,
  })

  if (!list.length) return pool.end()

  const update = R.partial(updatePool, [pool])
  const updates = await Promise.map(
    list,
    async row => {
      log.verbose({
        type: 'updating',
        rowId: row.id,
        url: row.evt.url,
      })
      let deets
      try {
        deets = await getDetails(row)
      } catch (e) {
        log.verbose({
          type: 'error getting details',
          error: e.message.substr(0, 25),
          row,
        })
        if (e.is404) {
          await update(false, undefined, row.id)
        }
        return false
      }

      if (deets.length > 1) {
        log.verbose({
          type: 'more than 1 event found',
          row,
          deets,
        })
        return false
      }

      let address = R.path(['location', 'address'])(deets[0])

      if (!address) {
        log.verbose({
          type: 'address is empty',
          row,
          deets,
        })
        try {
          address = await getPlace(deets[0])
        } catch (e) {
          log.verbose({
            type: 'error geocoding',
            address: address,
            error: e.message.substr(0, 25),
          })
          return false
        }
      }

      let geo
      try {
        geo = await geocode(address)
      } catch (e) {
        log.verbose({
          type: 'error geocoding',
          address: address,
          error: e.message.substr(0, 25),
        })
        return false
      }

      return update(deets[0], geo, row.id)
    },
    { concurrency: 5 },
  )

  const fails = updates.filter(i => !i).length
  log.info({
    type: 'details',
    fails,
    saves: updates.length - fails,
  })

  await pool.end()
}

export async function getDetails(row) {
  const { data } = await axios.get(row.evt.url)
  if (!data.match('DataObject type="event"')) {
    const err = new Error('event not found')
    err.is404 = true
    throw err
  }
  const { microdata } = WAE().parse(data.replace(/<br\s?\/?>/gi, '\n'))
  if (!microdata.Event) {
    throw new Error('microdata.Event is empty')
  }

  return microdata.Event
}

export async function geocode(address) {
  const params = {
    address,
    key: process.env.GOOGLE_API_KEY,
    region: 'ca',
  }
  const { data } = await axios.get(
    `https://maps.googleapis.com/maps/api/geocode/json?${qs(params)}`,
  )
  if (data.status !== 'OK') {
    throw new Error(
      `geocode status not OK: ${data.status}: ${data.error_message}`,
    )
  }
  return data
}

export async function getPlace(row) {
  const name = R.path(['location', 'name'])(row)
  if (!name) throw new Error('findplacefromtext no details.location.name found')
  const params = {
    key: process.env.GOOGLE_API_KEY,
    inputtype: 'textquery',
    fields: 'formatted_address',
    locationbias: '49.2827291,-123.1207375',
    input: name,
  }
  const { data } = await axios.get(
    `https://maps.googleapis.com/maps/api/place/findplacefromtext/json?${qs(
      params,
    )}`,
  )
  if (data.status !== 'OK') {
    throw new Error(
      `findplacefromtext status not OK: ${data.status}: ${data.error_message}`,
    )
  }
  return R.path(['candidates', '0', 'formatted_address'])(data)
}

export async function updatePool(pool, details, geo, id) {
  return pool.query(`update events set details=$1, geo=$2 where id=$3;`, [
    details,
    geo,
    id,
  ])
}
