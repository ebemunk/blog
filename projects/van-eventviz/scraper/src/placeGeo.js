import * as R from 'ramda'
import axios from 'axios'
import Promise from 'bluebird'

import { logger, qs } from './util'
import { getPool } from './db'

const log = logger({
  slack: false,
})

export default async function placeGeo(args, opts) {
  log.verbose({
    type: 'env',
    env: process.env,
  })
  const pool = getPool()
  const { rows: list } = await pool.query(
    `select * from events where
      geo='false'
      and details->'location'->'name' is not null
    ;`,
    // `select * from events where id='1160706';`,
  )
  log.verbose({
    type: 'found events where geo=false',
    count: list.length,
  })

  if (!list.length) return pool.end()

  const update = R.partial(updatePool, [pool])
  await Promise.map(
    list,
    async row => {
      log.verbose({
        type: 'updating',
        rowId: row.id,
        url: row.evt.url,
      })
      let place
      try {
        place = await getPlace(row)
      } catch (e) {
        log.verbose({
          type: 'error getting place',
          error: e.message, //.substr(0, 25),
          row,
        })
        return false
      }

      let geo
      try {
        geo = await geocode(place.candidates[0].formatted_address)
      } catch (e) {
        log.verbose({
          type: 'error geocoding',
          address: place.candidates[0].formatted_address,
          error: e.message, //.substr(0, 25),
        })
        return false
      }

      return update(geo, row.id)
    },
    { concurrency: 5 },
  )

  await pool.end()
}

export async function getPlace(row) {
  const name = R.path(['details', 'location', 'name'])(row)
  if (!name) throw new Error('findplacefromtext no details.location.name found')
  const params = {
    key: process.env.GOOGLE_API_KEY,
    inputtype: 'textquery',
    fields: 'formatted_address',
    locationbias: '49.2827291,-123.1207375',
    input: name,
  }
  const { data } = await axios.get(
    `https://maps.googleapis.com/maps/api/place/findplacefromtext/json?${qs(
      params,
    )}`,
  )
  if (data.status !== 'OK') {
    throw new Error(
      `findplacefromtext status not OK: ${data.status}: ${data.error_message}`,
    )
  }
  return data
}

export async function geocode(address) {
  const params = {
    address,
    key: process.env.GOOGLE_API_KEY,
    region: 'ca',
  }
  const { data } = await axios.get(
    `https://maps.googleapis.com/maps/api/geocode/json?${qs(params)}`,
  )
  if (data.status !== 'OK') {
    throw new Error(
      `geocode status not OK: ${data.status}: ${data.error_message}`,
    )
  }
  return data
}

export async function updatePool(pool, geo, id) {
  return pool.query(`update events set geo=$1 where id=$2;`, [geo, id])
}
