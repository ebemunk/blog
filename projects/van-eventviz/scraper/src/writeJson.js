import * as R from 'ramda'
import axios from 'axios'

import { logger } from './util'
import { getPool } from './db'

const log = logger({
  slack: {},
})

const dotPath = R.useWith(R.path, [R.split('.')])
const propsDotPath = R.useWith(R.ap, [R.map(dotPath), R.of])
const deepPick = paths => data => {
  return R.zipObj(paths, propsDotPath(paths)(data))
}

export default async function writeJson(args, opts) {
  log.verbose({
    type: 'env',
    env: process.env,
  })
  const pool = getPool()
  const { rows } = await pool.query(
    `
    select *
    from events where
      geo is not null
      and geo != 'false'::jsonb
      and (
        to_date(details->>'startDate', 'YYYY-MM-DD"T"HH24:MI:SS') > now()
        or to_date(details->>'endDate', 'YYYY-MM-DD"T"HH24:MI:SS') > now()
      )
    ;
    `,
  )
  const data = rows.map(row => ({
    id: row.id,
    tags: row.evt.tags,
    url: row.evt.url,
    ...(row.details &&
      deepPick([
        'name',
        'image',
        'endDate',
        'startDate',
        'description',
        'location.name',
      ])(row.details)),
    ...(row.geo &&
      deepPick(['geometry.location', 'formatted_address'])(row.geo.results[0])),
  }))
  const json = JSON.stringify(data)
  const jsonSize = json.length / (1024 * 1024)

  // require('fs').writeFileSync('../viz/data.json', JSON.stringify(data))
  await axios.put(
    'https://jsonblob.com/api/jsonBlob/6d894d47-0ca2-11e8-a2ea-5d19acf93d35',
    json,
    {
      headers: { 'Content-Type': 'application/json; charset=utf-8' },
    },
  )

  log.info({
    type: 'json upload',
    size: jsonSize,
  })

  await pool.end()
}
