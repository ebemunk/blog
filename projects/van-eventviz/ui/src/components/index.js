export Map from './Map'
export Marker from './Marker'
