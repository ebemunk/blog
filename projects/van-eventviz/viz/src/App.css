:global(body) {
  margin: 0;
  padding: 0;
  height: 100%;
  font-size: 12px;
}

:global(#root) {
  height: 100%;
  display: flex;
  flex-direction: column;
}

.wrap {
  display: flex;
  flex: 1;
}

.map {
  position: relative;
  flex: 1;
}

.side {
  flex-basis: 25%;
}
