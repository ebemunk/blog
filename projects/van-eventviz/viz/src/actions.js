import { get } from 'axios'

export const getEvents = () => async dispatch => {
  const { data: events } = await get(
    'https://jsonblob.com/api/jsonBlob/6d894d47-0ca2-11e8-a2ea-5d19acf93d35',
  )

  dispatch({
    type: 'EVENTS',
    payload: events,
  })
}

export const toggleTag = tag => ({
  type: 'TOGGLE_TAG',
  payload: tag,
})

export const toggleAll = () => ({
  type: 'TOGGLE_ALL',
})

export const toggleNone = () => ({
  type: 'TOGGLE_NONE',
})

export const selectDates = range => ({
  type: 'SELECT_DATES',
  payload: range,
})
