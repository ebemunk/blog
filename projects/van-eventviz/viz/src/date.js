import {
  startOfDay,
  endOfDay,
  startOfWeek as startOfWeekSun,
  endOfWeek as endOfWeekSun,
  endOfMonth,
  subDays,
  addDays,
  getMinutes,
  format as fmt,
  isSameDay,
} from 'date-fns'
import { compose, partialRight } from 'ramda'

const startOfWeek = partialRight(startOfWeekSun, [{ weekStartsOn: 1 }])
const endOfWeek = partialRight(endOfWeekSun, [{ weekStartsOn: 1 }])
const startOfSaturday = compose(
  startOfDay,
  partialRight(subDays, [1]),
  endOfWeek,
)

export const dayClickRange = ({ from, to }) => ({
  from: from ? startOfDay(from) : null,
  to: to ? endOfDay(to) : null,
})

export const todayRange = now => ({
  from: now,
  to: endOfDay(now),
})

export const thisWeekRange = now => ({
  from: now,
  to: endOfWeek(now),
})

export const thisWeekendRange = now => ({
  from: startOfSaturday(now),
  to: endOfWeek(now),
})

export const nextWeekRange = now => {
  const nextWeek = addDays(now, 7)
  return {
    from: startOfWeek(nextWeek),
    to: endOfWeek(nextWeek),
  }
}

export const thisMonthRange = now => ({
  from: now,
  to: endOfMonth(now),
})

export const format = date => {
  if (getMinutes(date) == 0) return fmt(date, 'ddd MMM DD @ ha')
  return fmt(date, 'ddd MMM DD @ h:ma')
}

export const formatTime = date =>
  getMinutes(date) === 0 ? fmt(date, 'ha') : fmt(date, 'h:ma')

export const formatEventDate = (start, end) => {
  if (start.getTime() == end.getTime()) return format(start)
  const endd = isSameDay(start, end) ? formatTime(end) : format(end)
  return `${format(start)} - ${endd}`
}
