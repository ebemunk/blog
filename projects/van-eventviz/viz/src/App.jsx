import React from 'react'

import Title from './components/Title'
import Map from './components/Map'
import Menu from './components/Menu'
import Filters from './components/Filters'
import List from './components/List'

import { todayRange } from './date'

import style from './App.css'

export class App extends React.Component {
  componentWillMount() {
    this.props.selectDates(todayRange(new Date()), 'today')
  }

  render() {
    return (
      <React.Fragment>
        <Title />
        <Map />
        <Menu />
        <Filters />
        <List />
      </React.Fragment>
    )
  }
}

import { connect } from 'react-redux'
import { selectDates } from './actions'

export default connect(null, {
  selectDates,
})(App)

// icons
//https://mapicons.mapsmarker.com/
