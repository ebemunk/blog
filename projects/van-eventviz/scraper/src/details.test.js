import axios from 'axios'
import Promise from 'bluebird'

import details, { getDetails, geocode } from './details'
import events from '../test/events.json'
import { readFile } from '../src/util'
import * as db from './db'

afterEach(() => {
  jest.restoreAllMocks()
})

describe('getDetails', () => {
  ;['1016021', '1016356', '1018066', '1018551'].map(id =>
    it(`should parse event ${id}`, async () => {
      jest
        .spyOn(axios, 'get')
        .mockImplementation(() =>
          Promise.props({ data: readFile(`./test/${id}.html`) }),
        )
      const event = await getDetails(events[0])
      expect(event).toMatchSnapshot()
    }),
  )
})

xdescribe('details', () => {
  it('should do it', async () => {
    jest.spyOn(db, 'getPool').mockImplementation(() => ({
      query: () =>
        Promise.resolve({
          rows: [
            {
              evt: { url: '' },
            },
          ],
        }),
      end: () => Promise.resolve(),
    }))
    jest.spyOn(axios, 'get').mockImplementation(
      () => Promise.props({ data: readFile(`./test/1018551.html`) }),
      // Promise.props({}),
    )

    await details()
  })
})

xdescribe('geocode', () => {
  it('does it', async () => {
    try {
      const res = await geocode('200 - 1131 Howe Street, Vancouver')
      console.log(res)
    } catch (e) {
      console.log(e)
    }
  })
})
