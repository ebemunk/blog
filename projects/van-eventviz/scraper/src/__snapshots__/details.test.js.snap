// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDetails should parse event 1016021 1`] = `
Array [
  Object {
    "@context": "https://schema.org/",
    "@type": "Event",
    "description": "Join the Mindchill gang for this 3hr, ‘all-levels’, workshop where we celebrate your inner ROCK STAR and explore creativity on the mat. Radical Authenticity is a journey into the freedom of a customised practice.

Each of us is unique, and each of us has our own experience both on and off the mat. Take your Yoga to new dimensions of personal growth and insight by tailoring the postures to suit your own body. Rather than using your body to ‘get into’ the postures, explore playing with the postures to ‘get into’ your body.

Unleash your hidden potential by honouring what is true…feel the luxurious bliss that comes from being authentically you!",
    "endDate": "2018-02-11T16:00:00-08:00",
    "image": Array [
      "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_standard/public/events/18/01/semperviva_georgiastraight_eventheaders_feb11_wescam.jpg?itok=i_eFnPMx",
      "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/18/01/semperviva_georgiastraight_eventheaders_feb11_wescam.jpg?itok=cQl3d_Zb",
    ],
    "location": Object {
      "@context": "http://schema.org/",
      "@type": "Place",
      "address": "1985 W. Broadway, Vancouver",
      "name": "Semperviva Yoga",
      "url": "/listings/venues/95191",
    },
    "name": "Radical Authenticity",
    "startDate": "2018-02-11T13:00:00-08:00",
  },
]
`;

exports[`getDetails should parse event 1016356 1`] = `
Array [
  Object {
    "@context": "https://schema.org/",
    "@type": "Event",
    "description": "Charles Condomine’s dinner party turns into mayhem when a seance led by Madame Arcati provokes the ghost of Elvira, his late first wife, and only Charles can hear her. Will Charles ever be free of Elvira? Or will his new wife, Ruth, also join her in haunting him too? It’s Madame Arcati to the rescue again! Deep Cove Stage Society's entry into the Theatre BC North Shore Zone Festival of Plays",
    "endDate": "2018-04-21T20:00:00-07:00",
    "location": Object {
      "@context": "http://schema.org/",
      "@type": "Place",
      "address": "4360 Gallant Ave., North Van",
      "name": "Deep Cove Shaw Theatre",
      "url": "/listings/venues/472716",
    },
    "name": "Blithe Spirit",
    "startDate": "2018-04-06T20:00:00-07:00",
  },
]
`;

exports[`getDetails should parse event 1018066 1`] = `
Array [
  Object {
    "@context": "https://schema.org/",
    "@type": "Event",
    "description": "The Eternal Tides is a momentous occasion—an adventure in beauty unlike anything seen before on the PuSh stage. In her Canadian debut, choreographer Lin Lee-Chen folds age-old rituals, customs and ceremonial rites into her radical vision, and the result is epic. The show is, above all, a tribute to water: the ocean that encloses the island of Taiwan, and the cycle of renewal it is part of. Using song, dance and striking stage design, Lin evokes the mythology of these elements. You might call it spiritual environmentalism—a tribute to nature as it was, and might be again.

With its large ensemble of dancers and musicians, the show has an intoxicating power. Lin carries us through a procession of vivid tableaux; the imagery evokes myth and human reality in equal measure. The costumes and sets are stunning: fabric and silver grass, red-tinted figures and shocks of white, composed with flair and delicacy. At once austere and sensual, intimate and monumental, this show will leave you awestruck.",
    "name": "The Eternal Tides | Legend Lin Dance Theatre (Taiwan)",
  },
]
`;

exports[`getDetails should parse event 1018551 1`] = `
Array [
  Object {
    "@context": "https://schema.org/",
    "@type": "Event",
    "description": "MEL GAUSDEN | GIRLS GONE WILD
BAU-XI VANCOUVER UPPER GALLERY | JANUARY 13 - 27
Opening Reception: Saturday, January 13, 2-4PM 
ARTIST IN ATTENDANCE

Canadian landscape painting has mythologized the great outdoors through the work of Tom Thomson, A.Y. Jackson and David Milne. As a genre historically dominated by male painters, emerging artist Mel Gausden marks a departure from the landscape tradition by centralizing the experience of women in nature, and her own in ﻿Girls Gone Wild, the artist's inaugural exhibition at Bau-Xi Vancouver.

Typically overlooked in traditional landscape painting, Gausden's canvases feature women as outdoors people: active participants who build fires, climb fences, and carry their own canoes. Her female subjects, often situated or leaving evidence of their trace, discover themselves within the natural environment in which they gather, explore and commune.

The visual culture of our contemporary digital age informs Gausden's unconventional palette. Neon silhouettes and brightly hued accents take reference from pre-set Instagram camera filters, fashion and consumer trends. The artist's manipulation of oil paint varies from sculptural— bending and piling thick layers of paint into matchsticks of textured impasto—to painterly— diluting paint until it drips with controlled precision— to sketchlike with her mark-making. For Gausden, the act of painting is intuitive as much as it is intentional; an exploratory act of personal and cultural identity which fuses past and present influences to create new possibilities for representation.

ABOUT THE ARTIST

Mel Gausden explores themes of personal and cultural identity through traditions of self-portraiture and landscape painting. Her interests lie in combining visual markers and signifiers of nostalgia and Canadiana with painterly gesture, mark-making and bold highlights. The artist applies neon silhouettes evocative of “instagram-esque” colour filters to highlight the inherent distortion of our own personal narratives. In the end, these visual narratives end up as hybrids, part-reality and part-dreamscape.

Mel Gausden earned her Bachelor’s of Arts at the Ontario College of Art and Design in 2009. She continues to live and work out of her studio in Toronto.",
    "endDate": "2018-01-27T14:00:00-08:00",
    "image": Array [
      "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_standard/public/events/18/01/gausden_ladybusiness-web.jpg?itok=mOe9zyj1",
      "//d2ciprw05cjhos.cloudfront.net/files/v3/styles/gs_small/public/events/18/01/gausden_ladybusiness-web.jpg?itok=PI5OHotM",
    ],
    "location": Object {
      "@context": "http://schema.org/",
      "@type": "Place",
      "address": "3045 Granville, Vancouver",
      "name": "Bau-Xi Gallery",
      "url": "/listings/venues/109236",
    },
    "name": "Girls Gone Wild",
    "startDate": "2018-01-13T14:00:00-08:00",
  },
]
`;
